step/step.go

```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
+		}
+	}()
+
+	if err := step.build(input); err != nil {
+		return fmt.Errorf("build docker image: %w", err)
+	}
+
+	// To make sure the cache is not growing indefinitely,
+	// we remove the original cache folder and move the new one to its place
+	if err := step.moveCacheFolder(dockerCacheFolderTemporary, dockerCacheFolder); err != nil {
+		return fmt.Errorf("move cache folder: %w", err)
+	}
+
+	return nil
+}
+
+func (step DockerBuildPushStep) destroyContainer(container string) error {
+	args := []string{
+		"buildx", "rm", "--force", container,
+	}
+	cmd := step.commandFactory.Create("docker", args, nil)
+	out, err := cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return fmt.Errorf("remove buildx instance %s: %w", out, err)
+	}
+	return nil
+}
+
+func (step DockerBuildPushStep) build(input Input) error {
+	stdout := NewLoggerWriter(step.logger)
+	defer stdout.Flush()
+
+	args := []string{
+		"buildx",
+		"build",
+	}
+
+	if input.BuildArg != "" {
+		for _, arg := range strings.Split(input.BuildArg, "\n") {
+			args = append(args, "--build-arg", arg)
+		}
+	}
+
+	switch {
+	case input.UseBitriseCache:
+		args = append(args, fmt.Sprintf("--cache-from=type=local,src=%s", dockerCacheFolder))
+		args = append(args, fmt.Sprintf("--cache-to=type=local,dest=%s,mode=max,compression=zstd", dockerCacheFolderTemporary))
+	case input.CacheFrom != "":
+		for _, cacheFrom := range strings.Split(input.CacheFrom, "\n") {
+			args = append(args, fmt.Sprintf("--cache-from=%s", cacheFrom))
+		}
+		fallthrough
+	case input.CacheTo != "":
+		for _, cacheTo := range strings.Split(input.CacheTo, "\n") {
+			args = append(args, fmt.Sprintf("--cache-to=%s", cacheTo))
+		}
+	}
+
+	if input.ExtraOptions != "" {
+		for _, option := range strings.Split(input.ExtraOptions, "\n") {
+			// This regex splits the string by spaces, but keeps quoted strings together
+			// Example --build-arg "-X main.version=1.0.0" will be split into --build-arg and "-X main.version=1.0.0"
+			r := regexp.MustCompile(`[^\s"']+|"([^"]*)"|'([^']*)`)
+			result := r.FindAllString(option, -1)
+
+			// Remove quotes from the strings
+			var options []string
+			for _, result := range result {
+				options = append(options, strings.ReplaceAll(result, "\"", ""))
+			}
+
+			args = append(args, options...)
+		}
+	}
+
+	if input.Push {
+		args = append(args, "--push")
+	} else {
+		// The --load parameter is used to load the image into the local docker daemon
+		// This is needed because the docker buildx build command will keep the result in cache only,
+		// preventing the use of the image in the same build
+		args = append(args, "--load")
+	}
+
+	for _, tag := range strings.Split(input.Tags, "\n") {
+		args = append(args, "--tag", tag)
+	}
+
+	args = append(args, []string{"-f", input.File, input.Context}...)
+
+	step.logger.Infof("$ docker %s", strings.Join(args, " "))
+
+	buildxCmd := step.commandFactory.Create("docker", args, &command.Opts{
```

Comment:
You can just pass `nil` as the `Opts` if you don't want to customize the default behavior (which connects stdout to stdout, stderr to stderr).

Reply:
We do want to customize, this stdout variable references our custom logwriter

Reply:
We removed the logger wrapper as stdout works too to reduce complexity.